downstairs/src/extent_inner_raw.rs

```diff
@@ -325,24 +325,42 @@ impl ExtentInner for RawInner {
         // operation atomic.
         self.set_flush_number(new_flush, new_gen)?;
 
-        // Now, we fsync to ensure data is flushed to disk.  It's okay to crash
-        // before this point, because setting the flush number is atomic.
-        cdt::extent__flush__file__start!(|| {
-            (job_id.get(), self.extent_number, 0)
-        });
-        if let Err(e) = self.file.sync_all() {
-            /*
-             * XXX Retry?  Mark extent as broken?
-             */
-            return Err(CrucibleError::IoError(format!(
-                "extent {}: fsync 1 failure: {e:?}",
-                self.extent_number,
-            )));
+        if cfg!(feature = "omicron-build") {
+            // no-op, FIOFFS will be called in region_flush
+        } else {
+            // Now, we fsync to ensure data is flushed to disk.  It's okay to crash
+            // before this point, because setting the flush number is atomic.
+            cdt::extent__flush__file__start!(|| {
+                (job_id.get(), self.extent_number, 0)
+            });
+
+            if let Err(e) = self.file.sync_all() {
+                /*
+                 * XXX Retry?  Mark extent as broken?
+                 */
+                return Err(CrucibleError::IoError(format!(
+                    "extent {}: fsync 1 failure: {e:?}",
+                    self.extent_number,
+                )));
+            }
+
+            cdt::extent__flush__file__done!(|| {
+                (job_id.get(), self.extent_number, 0)
+            });
         }
+
+        cdt::extent__flush__done!(|| { (job_id.get(), self.extent_number, 0) });
+
+        Ok(())
+    }
+
+    fn post_flush(
```

Comment:
It weirds me out that `Extent::flush` doesn't actually flush, depending on `omicron-build`.  I understand why that's the case, but would like to propose an alternative API.  I think we should break flushing behavior into three functions:

- `pre_flush` contains everything through `self.set_flush_number(..)` in the current implementation
- `flush_inner` implements the `else` branch, i.e. calling `self.file.sync_all` (unconditionally!)
- `post_flush` is the same as in this PR

Then, we can have a default implementation of `flush` on `trait ExtentInner` which calls these functions one after the other, meaning unit tests that want to flush a single extent can just call `flush()`.  Right now, those unit tests are no longer actually syncing the file to disk!

More importantly, the decision to skip calling `flush_inner` (i.e. just calling `pre_flush` + `post_flush`) is moved into the `Region::region_flush` implementation, which is right next to the FIOFFS call, which makes it more obvious what's going on.